### PR TITLE
NDEV-3370. Cache the simulation results between state.external_account() calls

### PR DESCRIPTION
**Pull request by artem-yazkov**

### CODE REVIEW

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11362303263/artifacts/2062465841)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   62894 |      0     |      0.207 |    0.00105732  |        0.001  |
| getAccountInfo                    |   57448 |      0     |      0.202 |    0.00202761  |        0.001  |
| getBlocks                         |   25407 |      0     |      0.068 |    0.000750384 |        0.001  |
| getBlock                          |   15161 |      0     |      0.153 |    0.00185298  |        0.002  |
| getMultipleAccounts               |   13485 |      0     |      0.153 |    0.00241528  |        0.001  |
| getLatestBlockhash                |   10761 |      0     |      0.153 |    0.00163461  |        0.001  |
| getTransaction                    |    7376 |      0     |      0.306 |    0.0132309   |        0.005  |
| getBlockTime                      |    6169 |      0     |      0.156 |    0.000770627 |        0      |
| sendTransaction                   |    5063 |      0     |      0.142 |    0.00854098  |        0.004  |
| getBalance                        |    1827 |      0     |      0.083 |    0.0021538   |        0.001  |
| getSignatureStatuses              |    1130 |      0     |      0.035 |    0.00183805  |        0.001  |
| getVersion                        |     595 |      0     |      0.043 |    0.00200336  |        0.001  |
| isBlockhashValid                  |     547 |      0     |      0.039 |    0.000714808 |        0      |
| getSignaturesForAddress           |      63 |      0     |      0.018 |    0.00314286  |        0.002  |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000605263 |        0.001  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.00084375  |        0.001  |
| getTokenAccountBalance            |       6 |      0     |      0.017 |    0.004       |        0.0015 |
| getHealth                         |       4 |      0     |      0.001 |    0.00025     |        0      |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |        0.001  |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |        0      |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>

File: evm_loader/program/src/executor/state.rs [hunk 1]
```diff
@@ -449,20 +453,39 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
             .collect::<Vec<_>>();
 
         if !metas.iter().any(|m| (m.pubkey == address) && m.is_writable) {
+            // account with given address don't changed by current actions
+            // put offset to cache
+            self.data.cache.borrow_mut().actions_offset = self.data.actions.len();
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor/state.rs]: Other accounts may be affected by current actions.
`.actions_offset` should only be changed after actions have been emulated.

File: evm_loader/program/src/executor/state.rs [hunk 2]
```diff
@@ -449,20 +453,39 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
             .collect::<Vec<_>>();
 
         if !metas.iter().any(|m| (m.pubkey == address) && m.is_writable) {
+            // account with given address don't changed by current actions
+            // put offset to cache
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor/state.rs]: Please use a spell checker when writing comments in English.
Also, in general, comments should explain why something is happening, not what.

File: evm_loader/program/src/executor/state.rs [hunk 3]
```diff
@@ -431,13 +433,15 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
         Ok(cache.block_timestamp)
     }
 
+    #[allow(clippy::too_many_lines)]
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor/state.rs]: You don't need it here.
If you think this function is too long, split it into several smaller.

File: evm_loader/program/src/executor/state.rs [hunk 4]
```diff
@@ -431,13 +433,15 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
...
     async fn external_account(&self, address: Pubkey) -> Result<OwnedAccountInfo> {
         self.touch_solana(address);
-
+        // collect accounts for actions that we don't processed yet
         let metas = self
             .data
             .actions
             .iter()
+            .skip(self.data.cache.borrow_mut().actions_offset)
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor/state.rs]: Please only use `self.data.cache.borrow_mut()` once per function.
`let mut cache = self.data.cache.borrow_mut();`

Clippy may display a warning about holding a reference over an 'await' point.
This is a false positive and can be suppressed.

File: evm_loader/program/src/executor/state.rs [hunk 5]
```diff
@@ -449,20 +453,39 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
...
         let mut accounts = BTreeMap::<Pubkey, OwnedAccountInfo>::new();
-
         for m in metas {
-            self.touch_solana(m.pubkey);
+            self.touch_solana(m.pubkey); // touch account whatever cached or not
 
-            let account = self.backend.clone_solana_account(&m.pubkey).await;
-            accounts.insert(m.pubkey, account);
+            // do we have account in the cash? if not - take it from backend
+            let mut use_cache = false;
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor/state.rs]:
```rust
let mut use_cache = false;
if let Some(account) = self.data.cache.borrow_mut().accounts.get(&m.pubkey) {
    accounts.insert(m.pubkey, account.clone());
    use_cache = true;
}
if !use_cache {
    let acc = self.backend.clone_solana_account(&m.pubkey).await;
    accounts.insert(m.pubkey, acc);
}
```

`else` statement exists. 
```rust
let account = if let Some(account) = cache.accounts.get(&m.pubkey) {
    account.clone()
} else {
    self.backend.clone_solana_account(&m.pubkey).await
};
accounts.insert(m.pubkey, account);
```

File: evm_loader/program/src/executor/state.rs [hunk 6]
```diff
@@ -506,6 +529,15 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
...
             }
         }
+        // put offset to cache
+        self.data.cache.borrow_mut().actions_offset = self.data.actions.len();
+        for (addr, acc) in &accounts {
+            self.data
+                .cache
+                .borrow_mut()
+                .accounts
+                .insert(*addr, acc.clone()); // put accounts to cache
```
Comment (anton-lisanin) [inline evm_loader/program/src/executor/state.rs]: Please avoid cloning an account here.

Comment (anton-lisanin) [reply evm_loader/program/src/executor/state.rs]: Also, you only need to update the cache with accounts that have been changed in emulated actions.

**Review by anton-lisanin — CHANGES_REQUESTED** [review]

**Comment by neonlabstech** [comment]
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/11445754501/artifacts/2084003402)

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   58649 |      0     |      0.28  |    0.000996334 |        0.001  |
| getAccountInfo                    |   55185 |      0     |      0.149 |    0.00179402  |        0      |
| getBlocks                         |   24366 |      0     |      0.045 |    0.000694656 |        0.001  |
| getBlock                          |   15739 |      0     |      0.035 |    0.00160588  |        0.001  |
| getMultipleAccounts               |   13698 |      0     |      0.173 |    0.00232041  |        0.001  |
| getLatestBlockhash                |   10781 |      0     |      0.077 |    0.00161562  |        0.001  |
| getTransaction                    |    7608 |      0     |      0.128 |    0.0125262   |        0.006  |
| getBlockTime                      |    6163 |      0     |      0.1   |    0.000591108 |        0      |
| sendTransaction                   |    5191 |      0     |      0.353 |    0.0099004   |        0.004  |
| getBalance                        |    1772 |      0     |      0.092 |    0.00178499  |        0.001  |
| getSignatureStatuses              |    1143 |      0     |      0.099 |    0.0015818   |        0.001  |
| getVersion                        |     612 |      0     |      0.097 |    0.00211111  |        0      |
| isBlockhashValid                  |     553 |      0     |      0.026 |    0.000643761 |        0      |
| getSignaturesForAddress           |      58 |      0     |      0.02  |    0.00196552  |        0.001  |
| simulateTransaction               |      38 |      0     |      0.001 |    0.000605263 |        0.001  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.000875    |        0.001  |
| getTokenAccountBalance            |      11 |      0     |      0.094 |    0.00890909  |        0      |
| getHealth                         |       3 |      0     |      0.001 |    0.000666667 |        0.001  |
| getFirstAvailableBlock            |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getClusterNodes                   |       1 |      0.097 |      0.097 |    0.097       |        0.097  |

</details>

**Review by anton-lisanin — APPROVED** [review]

**Pull request merged**